File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/service/DockerService.java
package org.iumotionlabs.hefesto.feature.portinfo.service;

import org.iumotionlabs.hefesto.feature.portinfo.model.DockerInfo;
import org.iumotionlabs.hefesto.feature.portinfo.model.DockerInfo.PortMapping;
import org.iumotionlabs.hefesto.feature.portinfo.model.PortBinding;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Service for Docker container integration and port mapping detection.
 */
public final class DockerService {

    // Pattern for docker port output: 0.0.0.0:8080->80/tcp
    private static final Pattern PORT_MAPPING_PATTERN = Pattern.compile(
        "(?:([\\d.]+|::)?:)?(\\d+)->(\\d+)/(tcp|udp)"
    );

    private Boolean dockerAvailable;

    /**
     * Checks if Docker is available on the system.
     */
    public boolean isDockerAvailable() {
        if (dockerAvailable == null) {
            dockerAvailable = checkDockerAvailable();
        }
        return dockerAvailable;
    }

    private boolean checkDockerAvailable() {
        try {
            Process process = new ProcessBuilder("docker", "info")
                .redirectErrorStream(true)
                .start();
            int exitCode = process.waitFor();
            return exitCode == 0;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Gets Docker info for a container by PID.
     */
    public Optional<DockerInfo> getContainerByPid(long pid) {
        if (!isDockerAvailable()) {
            return Optional.empty();
        }

        try {
            // First, find container ID by PID using docker top or cgroup
            Optional<String> containerId = findContainerIdByPid(pid);
            if (containerId.isEmpty()) {
                return Optional.empty();
            }

            return getContainerInfo(containerId.get());
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    /**
     * Gets Docker info for a specific container by ID or name.
     */
    public Optional<DockerInfo> getContainerInfo(String containerIdOrName) {
        if (!isDockerAvailable()) {
            return Optional.empty();
        }

        try {
            Process process = new ProcessBuilder(
                "docker", "inspect", "--format",
                "{{.Id}}|{{.Name}}|{{.Config.Image}}|{{.State.Status}}",
                containerIdOrName
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line = reader.readLine();
                if (line != null && !line.contains("Error")) {
                    String[] parts = line.split("\\|");
                    if (parts.length >= 4) {
                        String containerId = parts[0];
                        String name = parts[1].startsWith("/") ? parts[1].substring(1) : parts[1];
                        String image = parts[2];
                        String status = parts[3];

                        List<PortMapping> portMappings = getPortMappings(containerIdOrName);

                        return Optional.of(new DockerInfo(containerId, name, image, status, portMappings));
                    }
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Ignore
        }

        return Optional.empty();
    }

    /**
     * Gets all running containers with their port bindings.
     */
    public List<DockerInfo> listRunningContainers() {
        if (!isDockerAvailable()) {
            return List.of();
        }

        List<DockerInfo> containers = new ArrayList<>();

        try {
            Process process = new ProcessBuilder(
                "docker", "ps", "--format", "{{.ID}}|{{.Names}}|{{.Image}}|{{.Status}}"
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    String[] parts = line.split("\\|");
                    if (parts.length >= 4) {
                        String containerId = parts[0];
                        String name = parts[1];
                        String image = parts[2];
                        String status = parts[3];

                        List<PortMapping> portMappings = getPortMappings(containerId);
                        containers.add(new DockerInfo(containerId, name, image, status, portMappings));
                    }
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Ignore
        }

        return containers;
    }

    /**
     * Gets port mappings for a container.
     */
    public List<PortMapping> getPortMappings(String containerIdOrName) {
        List<PortMapping> mappings = new ArrayList<>();

        try {
            Process process = new ProcessBuilder(
                "docker", "port", containerIdOrName
            ).redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    // Format: 80/tcp -> 0.0.0.0:8080
                    parsePortLine(line).ifPresent(mappings::add);
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Ignore
        }

        return mappings;
    }

    private Optional<PortMapping> parsePortLine(String line) {
        // Format: 80/tcp -> 0.0.0.0:8080 or 80/tcp -> [::]:8080
        Pattern pattern = Pattern.compile("(\\d+)/(tcp|udp)\\s*->\\s*([\\d.:\\[\\]]+):(\\d+)");
        Matcher m = pattern.matcher(line);

        if (m.find()) {
            int containerPort = Integer.parseInt(m.group(1));
            String protocol = m.group(2);
            String hostIp = m.group(3).replace("[", "").replace("]", "");
            int hostPort = Integer.parseInt(m.group(4));

            return Optional.of(new PortMapping(hostPort, containerPort, protocol, hostIp));
        }

        return Optional.empty();
    }

    /**
     * Finds container ID by process PID.
     */
    private Optional<String> findContainerIdByPid(long pid) {
        // Method 1: Check cgroup (Linux)
        Optional<String> fromCgroup = findContainerIdFromCgroup(pid);
        if (fromCgroup.isPresent()) {
            return fromCgroup;
        }

        // Method 2: Check all containers' top
        return findContainerIdFromDockerTop(pid);
    }

    private Optional<String> findContainerIdFromCgroup(long pid) {
        try {
            Process process = new ProcessBuilder("cat", "/proc/" + pid + "/cgroup")
                .redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    // Look for docker container ID in cgroup path
                    Pattern pattern = Pattern.compile("docker[/-]([a-f0-9]{64})");
                    Matcher m = pattern.matcher(line);
                    if (m.find()) {
                        return Optional.of(m.group(1));
                    }
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Might not be Linux or no access
        }

        return Optional.empty();
    }

    private Optional<String> findContainerIdFromDockerTop(long pid) {
        try {
            // Get list of running containers
            Process process = new ProcessBuilder("docker", "ps", "-q")
                .redirectErrorStream(true).start();

            List<String> containerIds = new ArrayList<>();
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    containerIds.add(line.trim());
                }
            }
            process.waitFor();

            // Check each container's processes
            for (String containerId : containerIds) {
                if (containerHasPid(containerId, pid)) {
                    return Optional.of(containerId);
                }
            }
        } catch (Exception e) {
            // Ignore
        }

        return Optional.empty();
    }

    private boolean containerHasPid(String containerId, long pid) {
        try {
            Process process = new ProcessBuilder("docker", "top", containerId, "-o", "pid")
                .redirectErrorStream(true).start();

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    try {
                        if (Long.parseLong(line.trim()) == pid) {
                            return true;
                        }
                    } catch (NumberFormatException e) {
                        // Skip non-numeric lines (header)
                    }
                }
            }

            process.waitFor();
        } catch (Exception e) {
            // Ignore
        }

        return false;
    }

    /**
     * Checks if a port binding is from a Docker container.
     */
    public boolean isDockerBinding(PortBinding binding) {
        if (!isDockerAvailable()) {
            return false;
        }

        // Check if process name suggests Docker
        String name = binding.processName().toLowerCase();
        if (name.contains("docker") || name.contains("containerd")) {
            return true;
        }

        // Check by PID
        return findContainerIdByPid(binding.pid()).isPresent();
    }

    /**
     * Finds which container is using a specific host port.
     */
    public Optional<DockerInfo> findContainerByHostPort(int port) {
        List<DockerInfo> containers = listRunningContainers();

        for (DockerInfo container : containers) {
            for (PortMapping mapping : container.portMappings()) {
                if (mapping.hostPort() == port) {
                    return Optional.of(container);
                }
            }
        }

        return Optional.empty();
    }
}

File: hefesto-core/src/feature/portinfo/service/docker_service.rs
use std::process::Command;
use std::sync::OnceLock;

use hefesto_domain::portinfo::docker_info::{DockerInfo, PortMapping};
use hefesto_domain::portinfo::port_binding::PortBinding;

/// Cached Docker availability check.
static DOCKER_AVAILABLE: OnceLock<bool> = OnceLock::new();

/// Service for Docker container integration and port mapping detection.
///
/// Provides methods to detect running Docker containers, retrieve their port
/// mappings, and correlate them with system-level port bindings. Falls back
/// gracefully when Docker is not installed or the daemon is not running.
pub struct DockerService;

impl DockerService {
    /// Creates a new `DockerService`.
    pub fn new() -> Self {
        Self
    }

    /// Checks if Docker is available on the system.
    pub fn is_docker_available(&self) -> bool {
        *DOCKER_AVAILABLE.get_or_init(Self::check_docker_available)
    }

    /// Gets Docker info for a container by PID.
    pub fn get_container_by_pid(&self, pid: u32) -> Option<DockerInfo> {
        if !self.is_docker_available() {
            return None;
        }

        let container_id = self.find_container_id_by_pid(pid)?;
        self.get_container_info(&container_id)
    }

    /// Gets Docker info for a specific container by ID or name.
    pub fn get_container_info(&self, container_id_or_name: &str) -> Option<DockerInfo> {
        if !self.is_docker_available() {
            return None;
        }

        let output = Command::new("docker")
            .args([
                "inspect",
                "--format",
                "{{.Id}}|{{.Name}}|{{.Config.Image}}|{{.State.Status}}",
                container_id_or_name,
            ])
            .output()
            .ok()?;

        let line = String::from_utf8_lossy(&output.stdout);
        let line = line.trim();

        if line.is_empty() || line.contains("Error") {
            return None;
        }

        let parts: Vec<&str> = line.split('|').collect();
        if parts.len() < 4 {
            return None;
        }

        let container_id = parts[0].to_string();
        let name = parts[1].strip_prefix('/').unwrap_or(parts[1]).to_string();
        let image = parts[2].to_string();
        let status = parts[3].to_string();
        let port_mappings = self.get_port_mappings(container_id_or_name);

        Some(DockerInfo {
            container_id,
            container_name: name,
            image,
            status,
            port_mappings,
        })
    }

    /// Gets all running containers with their port bindings.
    pub fn list_running_containers(&self) -> Vec<DockerInfo> {
        if !self.is_docker_available() {
            return Vec::new();
        }

        let output = match Command::new("docker")
            .args(["ps", "--format", "{{.ID}}|{{.Names}}|{{.Image}}|{{.Status}}"])
            .output()
        {
            Ok(o) => o,
            Err(_) => return Vec::new(),
        };

        let stdout = String::from_utf8_lossy(&output.stdout);
        let mut containers = Vec::new();

        for line in stdout.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }

            let parts: Vec<&str> = line.split('|').collect();
            if parts.len() < 4 {
                continue;
            }

            let container_id = parts[0].to_string();
            let name = parts[1].to_string();
            let image = parts[2].to_string();
            let status = parts[3].to_string();
            let port_mappings = self.get_port_mappings(&container_id);

            containers.push(DockerInfo {
                container_id,
                container_name: name,
                image,
                status,
                port_mappings,
            });
        }

        containers
    }

    /// Gets port mappings for a container.
    pub fn get_port_mappings(&self, container_id_or_name: &str) -> Vec<PortMapping> {
        let output = match Command::new("docker")
            .args(["port", container_id_or_name])
            .output()
        {
            Ok(o) => o,
            Err(_) => return Vec::new(),
        };

        let stdout = String::from_utf8_lossy(&output.stdout);
        let mut mappings = Vec::new();

        for line in stdout.lines() {
            if let Some(mapping) = Self::parse_port_line(line) {
                mappings.push(mapping);
            }
        }

        mappings
    }

    /// Checks if a port binding is from a Docker container.
    pub fn is_docker_binding(&self, binding: &PortBinding) -> bool {
        if !self.is_docker_available() {
            return false;
        }

        // Check if process name suggests Docker
        let name = binding.process_name.to_lowercase();
        if name.contains("docker") || name.contains("containerd") {
            return true;
        }

        // Check by PID
        self.find_container_id_by_pid(binding.pid).is_some()
    }

    /// Finds which container is using a specific host port.
    pub fn find_container_by_host_port(&self, port: u16) -> Option<DockerInfo> {
        let containers = self.list_running_containers();

        for container in containers {
            for mapping in &container.port_mappings {
                if mapping.host_port == port {
                    return Some(container);
                }
            }
        }

        None
    }

    // ── Private helpers ────────────────────────────────────────────────────

    fn check_docker_available() -> bool {
        Command::new("docker")
            .arg("info")
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .status()
            .map(|s| s.success())
            .unwrap_or(false)
    }

    /// Parses a Docker port output line.
    /// Format: `80/tcp -> 0.0.0.0:8080` or `80/tcp -> [::]:8080`
    fn parse_port_line(line: &str) -> Option<PortMapping> {
        // Expected format: "80/tcp -> 0.0.0.0:8080"
        let parts: Vec<&str> = line.split("->").collect();
        if parts.len() != 2 {
            return None;
        }

        let container_part = parts[0].trim();
        let host_part = parts[1].trim();

        // Parse container port and protocol: "80/tcp"
        let container_parts: Vec<&str> = container_part.split('/').collect();
        if container_parts.len() != 2 {
            return None;
        }

        let container_port: u16 = container_parts[0].parse().ok()?;
        let protocol = container_parts[1].to_string();

        // Parse host address and port: "0.0.0.0:8080" or "[::]:8080"
        let host_part_clean = host_part.replace(['[', ']'], "");
        let last_colon = host_part_clean.rfind(':')?;

        let host_ip = host_part_clean[..last_colon].to_string();
        let host_port: u16 = host_part_clean[last_colon + 1..].parse().ok()?;

        Some(PortMapping {
            host_port,
            container_port,
            protocol,
            host_ip,
        })
    }

    fn find_container_id_by_pid(&self, pid: u32) -> Option<String> {
        // Method 1: Check cgroup (Linux)
        if let Some(id) = self.find_container_id_from_cgroup(pid) {
            return Some(id);
        }

        // Method 2: Check all containers' top
        self.find_container_id_from_docker_top(pid)
    }

    fn find_container_id_from_cgroup(&self, pid: u32) -> Option<String> {
        let cgroup_path = format!("/proc/{pid}/cgroup");
        let output = Command::new("cat")
            .arg(&cgroup_path)
            .output()
            .ok()?;

        let stdout = String::from_utf8_lossy(&output.stdout);
        let re = regex::Regex::new(r"docker[/-]([a-f0-9]{64})").ok()?;

        for line in stdout.lines() {
            if let Some(captures) = re.captures(line) {
                if let Some(id) = captures.get(1) {
                    return Some(id.as_str().to_string());
                }
            }
        }

        None
    }

    fn find_container_id_from_docker_top(&self, pid: u32) -> Option<String> {
        let output = Command::new("docker")
            .args(["ps", "-q"])
            .output()
            .ok()?;

        let stdout = String::from_utf8_lossy(&output.stdout);
        let container_ids: Vec<&str> = stdout.lines().map(|l| l.trim()).filter(|l| !l.is_empty()).collect();

        for container_id in container_ids {
            if self.container_has_pid(container_id, pid) {
                return Some(container_id.to_string());
            }
        }

        None
    }

    fn container_has_pid(&self, container_id: &str, pid: u32) -> bool {
        let output = match Command::new("docker")
            .args(["top", container_id, "-o", "pid"])
            .output()
        {
            Ok(o) => o,
            Err(_) => return false,
        };

        let stdout = String::from_utf8_lossy(&output.stdout);
        for line in stdout.lines() {
            if let Ok(line_pid) = line.trim().parse::<u32>() {
                if line_pid == pid {
                    return true;
                }
            }
        }

        false
    }
}

impl Default for DockerService {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_port_line_standard() {
        let mapping = DockerService::parse_port_line("80/tcp -> 0.0.0.0:8080");
        assert!(mapping.is_some());
        let m = mapping.unwrap();
        assert_eq!(m.host_port, 8080);
        assert_eq!(m.container_port, 80);
        assert_eq!(m.protocol, "tcp");
        assert_eq!(m.host_ip, "0.0.0.0");
    }

    #[test]
    fn test_parse_port_line_ipv6() {
        let mapping = DockerService::parse_port_line("80/tcp -> [::]:8080");
        assert!(mapping.is_some());
        let m = mapping.unwrap();
        assert_eq!(m.host_port, 8080);
        assert_eq!(m.host_ip, "::");
    }

    #[test]
    fn test_parse_port_line_invalid() {
        assert!(DockerService::parse_port_line("invalid").is_none());
        assert!(DockerService::parse_port_line("").is_none());
    }

    #[test]
    fn test_parse_port_line_udp() {
        let mapping = DockerService::parse_port_line("53/udp -> 0.0.0.0:5353");
        assert!(mapping.is_some());
        let m = mapping.unwrap();
        assert_eq!(m.protocol, "udp");
        assert_eq!(m.container_port, 53);
        assert_eq!(m.host_port, 5353);
    }
}
